test/index-intervals.spec.js

```diff
@@ -1,100 +1,142 @@
+    interval: MS_IN_DAY
+});
+
+function validate_indices(indices, interval) {
+    function _regex_for_interval(interval) {
+        switch (interval) {
+            case 'day':
+                return /day[0-9][0-9][0-9][0-9]\.[0-9][0-9]\.[0-9][0-9]/;
+            case 'week':
+                return /week[0-9][0-9][0-9][0-9]\.[0-9][0-9]?/;
+            case 'month':
+                return /month[0-9][0-9][0-9][0-9]\.[0-9][0-9]/;
+            case 'year':
+                return /year[0-9][0-9][0-9][0-9]/;
+            case 'none':
+                return /none/;
+            default:
+                throw new Error(`invalid interval ${interval}`);
+        }
+    }
+
+    var regex = _regex_for_interval(interval);
+    indices.forEach(function(index) {
+        expect(regex.test(index)).equal(true);
     });
+}
 
-    function check(interval, suffix) {
-        var index = suffix ? interval + '*' : 'none';
-        return test_utils.write(points, {index: index, indexInterval: interval})
-            .then(function(result) {
-                expect(result.errors).deep.equal([]);
-                return test_utils.verify_import(points, 'local', index);
+modes.forEach(function(mode) {
+    describe('index intervals -- ' + mode, function() {
+        afterEach(function() {
+            var id = test_utils.test_id;
+            return test_utils.clear_data(mode, `${id}*`);
+        });
+
+        function check(interval) {
+            var suffix = interval !== 'none' ? interval + '*' : interval;
+            var index = test_utils.test_id + suffix;
+            return Promise.map(points, function(pt, n) {
+                return test_utils.write([pt], {index: index, indexInterval: interval, id: mode})
+                    .then(function(result) {
+                        expect(result.errors).deep.equal([]);
+                    });
+            }, {concurrency: 5})
+            .then(function(results) {
+                return test_utils.verify_import(points, mode, index);
             })
             .then(function() {
                 return test_utils.list_indices();
             })
             .then(function(indices) {
-                var week_indices = indices.filter(function(index) {
-                    return index.substring(index.length - suffix.length) === suffix;
-                });
-
-                expect(week_indices.length).at.least(1);
-                var start = '2014-09-17T14:13:42.000Z';
-                var end = '2014-10-17T14:13:42.000Z';
-                return test_utils.read({from: start, to: end, index: index, indexInterval: interval});
+                validate_indices(indices, interval);
+                var from = new Date(points[0].time).toISOString();
+                var to_ms = new Date(_.last(points).time).getTime();
+                var to = new Date(to_ms + 1).toISOString();
+                return test_utils.read({id: mode, from: from, to: to, index: index, indexInterval: interval});
```

Comment:
So this reads exactly from the index that we put data into. Let's have tests somewhere (if we do already, great!) that read cross-interval, like this (for each of the interval types: day, week, month, year):
a) read from a to b where no points had been written and no indexes created between a and b, get 0 points back;
b) read from a to b which covers 3 intervals: first had no data (index doesn't exist), second has data, third again has no data. The point is to make sure we don't barf reading from indices that don't exist;
c) read from a to b which covers 3 intervals, where all 3 have data. This is interesting specifically where indices cross over the month and year boundary. The point is to make sure that we correctly form the list of indices to hit in the url.